phantun/src/utils.rs

```diff
@@ -28,11 +29,81 @@ pub fn new_udp_reuseport(local_addr: SocketAddr) -> UdpSocket {
     // from tokio-rs/mio/blob/master/src/sys/unix/net.rs
     udp_sock.set_cloexec(true).unwrap();
     udp_sock.set_nonblocking(true).unwrap();
+
+    // enable IP_PKTINFO/IPV6_PKTINFO delivery so we know the destination address of incoming
+    // packets
+    if local_addr.is_ipv4() {
+        nix::sys::socket::setsockopt(&udp_sock, nix::sys::socket::sockopt::Ipv4PacketInfo, &true)
+            .unwrap();
+    } else {
+        nix::sys::socket::setsockopt(
+            &udp_sock,
+            nix::sys::socket::sockopt::Ipv6RecvPacketInfo,
+            &true,
+        )
+        .unwrap();
+    }
+
     udp_sock.bind(&socket2::SockAddr::from(local_addr)).unwrap();
     let udp_sock: std::net::UdpSocket = udp_sock.into();
     udp_sock.try_into().unwrap()
 }
 
+pub async fn udp_recv_pktinfo(
+    sock: &UdpSocket,
+    buf: &mut [u8],
+) -> std::io::Result<(usize, std::net::SocketAddr, std::net::IpAddr)> {
+    use std::os::unix::io::AsRawFd;
+    use tokio::io::Interest;
+
+    sock.async_io(Interest::READABLE, || {
+        // FIXME this is somewhat excessive, we actually need only
+        // max(sizeof(in_pktinfo), sizeof(in6_pktinfo))
+        let mut control_buffer = nix::cmsg_space!(nix::libc::in_pktinfo, nix::libc::in6_pktinfo);
```

Comment:
```suggestion
        // sometimes we might only need sizeof(in_pktinfo),
        // but it shouldn't hurt to just always allocate
        // sizeof(in6_pktinfo) here to save a syscall that
        // determines the address type of `sock`
        let mut control_buffer = nix::cmsg_space!(nix::libc::in6_pktinfo);
```

Would this be sufficient? We don't actually needs a buffer that is `sizeof(in_pktinfo) + sizeof(in6_pktinfo)` as you mentioned.

Reply:
Assuming `in6_pktinfo` is larger than `in_pktinfo`, yes. Which I think is true, but that's not immediately obvious.

What do you think about this alternative?
```suggestion
        use nix::sys::socket::cmsg_space;
        let control_buffer_size = std::cmp::max(
            cmsg_space::<nix::libc::in_pktinfo>(),
            cmsg_space::<nix::libc::in6_pktinfo>(),
        );
        let mut control_buffer = vec![0u8; control_buffer_size];
```